src/rules/no-missing-end-of-source-newline/index.js

```diff
@@ -15,16 +15,15 @@ export default function (actual) {
     const validOptions = validateOptions(result, ruleName, { actual })
     if (!validOptions) { return }
 
-    const sourceCss = root.toString()
-    if (sourceCss === "") { return }
-    if (sourceCss.slice(-1) !== "\n") {
-      report({
-        message: messages.rejected,
-        node: root,
-        index: root.toString().length - 1,
-        result,
-        ruleName,
-      })
-    }
+    const sourceCss = root.source.input.css
+    if (sourceCss === "" || sourceCss.slice(-1) === "\n") { return }
```

Comment:
@davidtheclark https://github.com/stylelint/stylelint/pull/1958 fix bug, i only add tests and speedup rule